plugin/session_buffer.py

```diff
@@ -329,23 +330,28 @@ def purge_changes_async(self, view: sublime.View) -> None:
         finally:
             self._pending_changes = None
         self.session.notify_plugin_on_session_buffer_change(self)
-        sublime.set_timeout_async(lambda: self._on_after_change_async(view, version))
+        sublime.set_timeout_async(lambda: self._on_after_change_async(view, version, suppress_requests))
```

Comment:
instead of passing suppress_requests,
can we just do an early return one line above:
```py
if suppress_requests:
   return
sublime.set_timeout_async(lambda: self._on_after_change_async(view, version))
```

I am wondering if we can avoid the `suppress_requests` entirely, 
by relying on the `self.opened` property?

```diff
        if self.opened and self.should_notify_did_close():
+           self.opened = False
+           self.purge_changes_async(view)
-           self.purge_changes_async(view, suppress_requests=True)
            self.session.send_notification(did_close(uri=self._last_known_uri))
-           self.opened = False
```

```diff
    def _on_after_change_async(self, view: sublime.View, version: int, suppress_requests: bool = False) -> None:
        if self._is_saving:
            self._has_changed_during_save = True
            return
       if not self.opened:
            return
```

Reply:
if we can't rely on self.opened, 
can we than use view.is_valid() to not send requests for a closed view?

Reply:
if I understood correctly suppress_requests is only True when we send the didClose notification
to not send requests for a view that is closing(or closed, didn't test to be 100% sure about this statement)

So it kind of make sense to rely on self.opened or view.is_valid().

Reply:
1. `view.is_valid()`:

    I think this is not possible to use. I use this test file to simulate the EventListener callbacks from ST that LSP uses:
    ```python
    import sublime
    import sublime_plugin
    
    
    class TestTextChangeListener(sublime_plugin.TextChangeListener):
    
        def on_text_changed(self, changes):
            print('on_text_changed')
    
    
    class TestViewEventListener(sublime_plugin.ViewEventListener):
    
        def on_post_save_async(self):
            print('on_post_save_async')
    
        def on_close(self):
            print('on_close', self.view.is_valid())
            sublime.set_timeout_async(lambda: print('async after close', self.view.is_valid()))
    ```
    When I close a view, I get this result:
    ```
    on_close True
    async after close True
    ```
    The view is still reported as valid here, and therefore we can't use it as a check whether to suppress the requests. In the real code, it will go through various other functions from DocumentSyncListener and SessionView before calling `purge_changes_async`. If I see it right, this is the line which puts in on the async thread: https://github.com/sublimelsp/LSP/blob/550d5910a14e7a2943d4aacd1f58f6049f3917d6/plugin/documents.py#L1016
    So it could mean that the result of `view.is_valid()` might even depend on how fast all the functions run and how busy the async thread is.

2. `self.opened`:
  
    I think this should work with your suggested diff. But it's a bit subjective which solution is the better one. If it checks the `self.opened` property, then the code implicitly relies on it being set to `False` before the function is called. If we forget about that and in a few years change the order of the lines again for some reasons, it might not work correctly anymore. So I think passing the explicit `suppress_requests` parameter would be the better alternative and makes the code easier to understand.

Reply:
The approach with `suppress_requests` is the most explicit.
 suppress_requests is effectively used just in `_on_after_change_async`
but we also need to pollute the purge_changes_async with an additional argument
`def purge_changes_async(self, view: sublime.View, suppress_requests: bool = False) -> None:`.

The first time I sat the suppress_requests in purge_changes_async,
it wasn't that clear what requests will be suppressed.
I thought that it would prevent sending the didChange **notifications**... but I was wrong.


I would be fine with `suppress_requests` or with `self.opened`... So I will resolve this thread.